MusicKit/Utilities/ScorefileUtils/playscore.tproj/playscore.1
.TH PLAYSCORE 1 "12 Mar 2001" "The MusicKit Project"
.UC 4
.SH NAME
playscore \- MusicKit scorefile playing utility
.SH SYNOPSIS
playscore [-fqdpva] [-rtcwsn <arg>] scorefile 
.SH DESCRIPTION
Plays the specified scorefile on the DSP and MIDI.  

If the scorefile
isn't found in the current directory and it isn't given as a full path
name, the following directories are searched on OpenStep 4.X machines:
.BL -enum
.IT ~/Library/MusicKit/Music/Scores
.IT /LocalLibrary/MusicKit/Music/Scores 
.IT /NextLibrary/MusicKit/Music/Scores
.EL

On MacOS X machines the following directories are searched:
.BL
.IT ~/Library/MusicKit/Music/Scores
.IT /Local/Library/MusicKit/Music/Scores 
.IT /Network/Library/MusicKit/Music/Scores
.IT /System/Library/MusicKit/Music/Scores
.EL

The actual search list of directories used will be listed in the usage message.
.SH OPTIONS
.TP
.B \-a
All DSPs; use all available DSP cards (Intel only).  This flag should not be
combined with \-c, \-w or \-s.  Instead, the serial port device should be set
with the Music Kit Preferences panel.
.TP
.B \-c <soundfile>
Write as commands; the DSP commands are written to the named
soundfile.  The '.snd' extension is automatically appended to the file
name. This mode is largely obsolete, as DSP commands files are not
supported on Intel hardware.
.TP
.B \-d
Debug; allocate space (on the DSP) for the DSP debugger.
.TP
.B \-f
Fast; normally, playscore reads the entire scorefile before playing;
with the f option, notes are played as soon as they are read from the
file.  This reduces start up time, but can also cause playback to be
less reliable.
.TP
.B \-n <dsp number>
Number of the DSP; specifies that the DSP<n> should be used as the default DSP.
Normally, playscore uses DSP0.   This flag should not be combined with \-a.
.TP
.B \-p
Persistent; waits for DSP if it's in use.
.TP
.B \-q
Quiet; error and warning messages are suppressed.
.TP
.B \-r <repeatCount>
Repeat; play the file <repeatCount> times.
.TP
.B \-s <deviceName>
Serial sound; Specifies that sound should be sent out the DSP serial port to the
specified device.  This flag is largely obsolete, thanks to the 
Music Kit Preferences panel.  Devices supported are SSAD64x (Singular Solutions),
StealthDAI2400 and GENERIC. In the case of Intel DSP cards that do not have
NeXT-compatible DSP ports, the serial port device is automatically selected.
An example of such a card is the TurtleBeach Multisound.
.TP
.B \-t <arg>
Trace; set Music Kit trace bit to <arg>.  Bits may be added.
The meanings of the bits are as follows: 
1, Orchestra allocation information; 
2, App params printed when first encountered;
4, Music Kit DSP messages;
8, MIDI in/out/time warnings;
16, SynthInstrument preemptions msgs;
32, SynthInstrument messages;
64, SynthPatch library messages;
128, UnitGenerator library messages;
256, Conductor time messages;
512, print arrays loaded to DSP.
.TP
.B \-v
Verbose; print DSP library error and status messages.
.TP
.B \-w <soundfile>
Write as samples; the DSP sample output is written to the named
soundfile.  The '.snd' extension is automatically appended to the file
name.
.SH "INSTRUMENTS"
The playscore program has access to the following instruments: 
.TP
.B Wave1i
Wave table synthesis.
.TP
.B Wave1vi
Wave table synthesis with vibrato.
.TP
.B DBWave1vi
Wave table synthesis with vibrato and access to the timbre data base.
.TP
.B DBWave2vi
Like DBWave1vi, but with interpolation between two wave tables.
.TP 
.B Pluck 
Plucked string synthesis.
.TP
.B Fm1i
Frequency modulation synthesis with arbitrary waveform.
.TP
.B Fm1vi
Frequency modulation synthesis with arbitrary waveform and vibrato.
.TP
.B DBFm1vi
Frequency modulation with vibrato and access to the timbre data base.
.TP
.B Fm2pvi
Frequency modulation with two parallel modulators and vibrato.
.TP
.B Fm2cvi
Frequency modulation synthesis with two cascaded modulators and vibrato.
.TP
.B Fm2pnvi
Like Fm2pvi, but with an additional noise modulator.
.TP
.B Fm2cnvi
Like Fm2cvi, but with an additional noise modulator.
.TP
.B Wave1, Wave1v, DBWave1v, Fm1v, Fm1
These are like the corresponding patches described above, but they use
a non\-interpolating oscillator.  This allows more simultaneous
voices, but at lower sound\-quality.
.TP
.B Simp
The simplest instrument:  a  single, non\-interpolating sine wave oscillator. 
.TP 
.B Waveshape
Waveshaping (non-linear distortion) synthesis.
.TP
.B midi0
MIDI output on the default MIDI device, MIDI0, 
which is serial port A on NeXT hardware. On 
Intel hardware, MIDI0 is set via the Music Kit Preferences panel.
"midi" is a synonym for "midi0"
.TP
.B midi1
MIDI output on the MIDI1 device, which is serial port B on NeXT hardware.
.SH "THE SCOREFILE INFO STATEMENT"
playscore decides which instrument to use, as well as other
configuration information, based on the info statements in the
scorefile.  The following scorefile info statement parameters are
recognized:
.TP 
.B headroom         
Headroom is a fractional number in the range -1 <= f <= 1 that
determines how DSP resources are allocated.  As the headroom is
decreased, the DSP becomes "bigger" but at the cost of decreased
reliability (generally heard as degraded rhythmic integrity).
Increasing the headroom makes synthesis more reliable, but you may
decrease the number of simultaneous voices (the DSP becomes
"smaller").  The default is 0.
.TP
.B samplingRate     
This should be 44100 or 22050. The default is 44100.
.TP
.B alternativeSamplingRate     
If present, this is used as the sampling rate for Turtle Beach cards,
which have slower DSP memory. The alternativeSamplingRate is a simple
way to provide a lower sampling rate for use with such cards. 
.TP
.B tempo
The tempo is in beats per minute; the default is 60.0.
.TP
.B midiOffset
This is a time offset in seconds for notes that are sent to an
external MIDI synthesizer.  It's used to adjust synchronization
between MIDI and the DSP.  Its value can be negative (shift MIDI notes
ahead with respect to DSP synthesis) or positive (shift MIDI notes
behind with respect to the DSP).
.SH "THE PART INFO STATEMENT"
Part info statements provide information that's applied to a
particular part in the scorefile. The following part info statement
parameters are recognized:
.TP 
.B synthPatch       
This sets the instrument for the part; you must quote the instrument
name, as taken from the instrument list given above.  For example,
.IP
synthPatch:"Pluck"
.TP 
.B synthPatchCount  
This is the number of preallocated instruments (simultaneous voices);
if it's omitted, the instruments are allocated as they're needed.  The
SynthPatch count is significant for DSP instruments only.
.TP 
.B midiChan  
This is the MIDI channel on which the part's notes are sent.  The
value must be an integer between 1 and 16, inclusive.  The MIDI
channel is significant for MIDI instruments only.
.TP 
.B orchestraIndex  
If you specify the \-a switch, the orchestraIndex part info parameter is used
to determine on which DSP to play the part. The value of this parameter is
0 for DSP0, 1 for DSP1, etc.  You set the value of these DSP variables from
the Music Kit Preferences panel.
.SH FILES
Example scorefiles are found in /Local/Library/MusicKit/Music/Scores.
You may copy these files to your directory and modify them with a Music Kit program
or a text editor.
There are two types of files. Files of the form Examp* or *Examp* are 
tutorial examples which illustrate the use of the software instruments.
Other files, such as 
.B Emma.playscore
and 
.B Gamelan.playscore
are short demo pieces.
.SH "SEE ALSO"
.B ScorePlayer.app
is an application very similar to playscore. However, currently,
playscore is the only one of the two that can handle multiple
simultaneous DSPs.  See the Music Kit documentation on
/Local/Library/Documentation/MusicKit for general information about
scorefiles, DSP software instruments, and MIDI.
.SH DIAGNOSTICS
Exit status is 0 on success, 1 if there is a problem opening the DSP
or parsing the file.
.SH BUGS
Large chords (exactly simultaneous notes) may cause timing problems.
It's better to separate them by some small amount of time.  This is
particularly true when doing wave table synthesis.  Frequent changes
of wave tables can cause the music to become uneven.  Very fast
musical passages may also be problematic.  In some cases, putting a
one-beat rest at the start of the scorefile may be helpful.  Playback
may become unreliable if your computer is heavily loaded with other
tasks.  For scores that are too dense or complex to play in real time,
try using the \-c or \-w flag and then play the resulting soundfile.
midiOffset shouldn't be necessary, but it is, for some reason.
.SH COPYRIGHT
Portions Copyright (c) 1992, NeXT Computer, Inc.
Portions Copyright (c) 1994, Stanford University.
Portions Copyright (c) 2000-2001, The MusicKit Project


MusicKit/Applications/PianoRoll/PlayScore.m
/* $Id$
  Plays scorefile in background. -- David Jaffe 
*/

#import <Foundation/Foundation.h>
#import <AppKit/AppKit.h>
#import <MusicKit/MusicKit.h>
#import "PlayScore.h"

@implementation PlayScore:NSObject

static NSMutableArray *synthInstruments;
static MKScorePerformer *scorePerformer;
static MKOrchestra *theOrch;
static double samplingRate = 22050;
static double headroom = .1;
static double initialTempo;

static BOOL userCancelFileRead = NO;

static void handleMKError(NSString *msg)
{
    if (![MKConductor inPerformance]) {
	if (!NSRunAlertPanel(@"PianoRoll", msg, @"OK", @"Cancel", nil, NULL)) {
	  MKSetScorefileParseErrorAbort(0);
	  userCancelFileRead = YES;         /* A kludge for now. */
      }
    }
    else {
	NSLog(msg);
    }
}

-(BOOL)isPlaying
{
    return (BOOL)(theOrch && [theOrch deviceStatus] == MK_devRunning);
}

- (void)setUpPlay: (MKScore *) scoreObj
{
   /* Could keep these around, in repeat-play cases: */ 
    id scoreInfo;
    if ([self isPlaying])
    	[self stop];
    samplingRate = 22050;
    headroom = .1;
//    MKSetTrace(1023);
    [[MKConductor defaultConductor] setTempo:initialTempo = 60];
    scoreInfo = [scoreObj infoNote];
    if (scoreInfo) { /* Configure performance as specified in info. */ 
	if ([scoreInfo isParPresent:MK_headroom])
            headroom = [scoreInfo parAsDouble:MK_headroom];	  
	if ([scoreInfo isParPresent:MK_samplingRate]) {
	    samplingRate = [scoreInfo parAsDouble:MK_samplingRate];
	    if (!((samplingRate == 44100.0) || (samplingRate == 22050.0))) 
		NSRunAlertPanel(@"ScorePlayer", @"Sampling rate must be 44100 or 22050.\n", @"OK", nil, nil);
	}
	if ([scoreInfo isParPresent:MK_tempo]) {
	    initialTempo = [scoreInfo parAsDouble:MK_tempo];
        [[MKConductor defaultConductor] setTempo:initialTempo];
	} 
    } 
    [scorePerformer release];
    scorePerformer = nil;  // Be Wary of this LMS
    [synthInstruments removeAllObjects];
    [synthInstruments release]; 
    synthInstruments = nil;  // Be wary of this LMS
}

- (BOOL) play:scoreObj
{
    int partCount,synthPatchCount,voices,i;
    NSString *className;
    id partPerformers,synthPatchClass,partPerformer,partInfo,anIns,aPart;

    if ([self isPlaying])
    	[self stop];
    theOrch = [MKOrchestra newOnDSP:0]; /* A noop if it exists */
//    [theOrch setHeadroom:headroom];    /* Must be reset for each play */ 
    [theOrch setSamplingRate:samplingRate];
    if (![theOrch open]) {
	NSRunAlertPanel(@"ScorePlayer", @"Can't open DSP. Perhaps another application has it.", @"OK", nil, nil);
	return NO;
    }
    scorePerformer = [MKScorePerformer new];
    [scorePerformer setScore:scoreObj];
    [scorePerformer activate]; 
    partPerformers = [scorePerformer partPerformers];
    partCount = [partPerformers count];
    synthInstruments = [[NSMutableArray alloc] init];
    for (i = 0; i < partCount; i++) {
	partPerformer = [partPerformers objectAtIndex:i];
	aPart = [partPerformer part]; 
	partInfo = [aPart infoNote];      
	if ((!partInfo) || ![partInfo isParPresent:MK_synthPatch]) {
	    if (!NSRunAlertPanel(@"ScorePlayer", 
                [NSString stringWithFormat: @"%@ info missing.\n", MKGetObjectName(aPart)], @"Continue", @"Cancel", nil)) 
	      return NO;
	    continue;
	}		
	className = [partInfo parAsStringNoCopy:MK_synthPatch];
        synthPatchClass = [MKSynthPatch findPatchClass:className];
        
	if (!synthPatchClass) {         /* Class not loaded in program? */ 
	    if (!NSRunAlertPanel(@"ScorePlayer", 
                [NSString stringWithFormat: @"This scorefile calls for a synthesis instrument (%@) that isn't available in this application.\n", className],
                     @"Continue", @"Cancel", nil))
	      return NO;
	    /* We would prefer to do dynamic loading here. */
	    continue;
	}
	anIns = [MKSynthInstrument new];      
	[synthInstruments addObject:anIns];
	[[partPerformer noteSender] connect:[anIns noteReceiver]];
	[anIns setSynthPatchClass:synthPatchClass];
	if (![partInfo isParPresent:MK_synthPatchCount])
	  continue;         
	voices = [partInfo parAsInt:MK_synthPatchCount];
	synthPatchCount = 
	  [anIns setSynthPatchCount:voices patchTemplate:
	   [synthPatchClass patchTemplateFor:partInfo]];
        [anIns release]; /* since retain is now held in synthInstruments array! */
	if (synthPatchCount < voices) {
	    if (!NSRunAlertPanel(@"ScorePlayer", 
                [NSString stringWithFormat: @"Could only allocate %d instead of %d %@s for %@\n",
		    synthPatchCount, voices, className, MKGetObjectName(aPart)], 
                    @"Continue", @"Cancel", nil))
	      return NO;
	}
    }
//    [partPerformers release];
    MKSetDeltaT(1.0);
    [MKConductor setClocked:YES];     
//    [MKOrchestra setTimed:YES];
    [MKConductor afterPerformanceSel:@selector(close) to:theOrch argCount:0];
//    [MKConductor afterPerformanceSel:@selector(hello) to:self argCount:0];

    [theOrch run];
    [MKConductor startPerformance];
    return YES; 
}

- init
{
    static int inited = 0;
    [super init];
    if (inited++)
        return self;
    [MKConductor setThreadPriority:1.0];
    [MKConductor useSeparateThread:YES];
    MKSetErrorProc(handleMKError);

    return self;
}
- (void)dealloc
{
    [scorePerformer release];
    [synthInstruments removeAllObjects];
    [synthInstruments release];
    [theOrch release];
}

- stop
{
    [MKConductor lockPerformance];
//    [theOrch abort];
    [MKConductor finishPerformance];
    [MKConductor unlockPerformance];
    [theOrch close];
    return self;
}

@end

